xds/src/main/java/com/linecorp/armeria/xds/client/endpoint/EndpointsPool.java

```diff
@@ -0,0 +1,100 @@
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.xds.client.endpoint;
+
+import static com.linecorp.armeria.internal.client.endpoint.RampingUpKeys.createdAtNanos;
+import static com.linecorp.armeria.internal.client.endpoint.RampingUpKeys.hasCreatedAtNanos;
+import static com.linecorp.armeria.internal.client.endpoint.RampingUpKeys.withCreatedAtNanos;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.client.endpoint.EndpointGroup;
+import com.linecorp.armeria.common.util.AsyncCloseable;
+import com.linecorp.armeria.xds.ClusterSnapshot;
+
+import io.netty.util.concurrent.EventExecutor;
+
+final class EndpointsPool implements AsyncCloseable {
+
+    private EndpointGroup delegate = EndpointGroup.of();
+    private Map<Endpoint, Long> createdTimestamps = ImmutableMap.of();
+    private final ClusterEntry clusterEntry;
+    private final EventExecutor eventExecutor;
+    private Consumer<List<Endpoint>> listener = ignored -> {};
+
+    EndpointsPool(ClusterEntry clusterEntry, EventExecutor eventExecutor) {
+        this.clusterEntry = clusterEntry;
+        this.eventExecutor = eventExecutor;
+    }
+
+    void updateClusterSnapshot(ClusterSnapshot newSnapshot) {
+        // clean up the old endpoint and listener
+        delegate.removeListener(listener);
+        delegate.closeAsync();
+
+        // set the new endpoint and listener
+        delegate = XdsEndpointUtil.convertEndpointGroup(newSnapshot);
+        listener = endpoints -> eventExecutor.execute(
+                () -> attachTimestampsAndDelegate(newSnapshot, endpoints));
+        delegate.addListener(listener, true);
+    }
+
+    private void attachTimestampsAndDelegate(ClusterSnapshot clusterSnapshot, List<Endpoint> endpoints) {
+        final long defaultTimestamp = System.nanoTime();
+        final ImmutableMap.Builder<Endpoint, Long> timestampsBuilder = ImmutableMap.builder();
+        final ImmutableList.Builder<Endpoint> endpointsBuilder = ImmutableList.builder();
+        for (Endpoint endpoint: endpoints) {
+            final Endpoint endpointWithTimestamp;
+            final long timestamp;
+            if (hasCreatedAtNanos(endpoint)) {
+                timestamp = createdAtNanos(endpoint);
+            } else {
+                timestamp = createdTimestamps.getOrDefault(endpoint, defaultTimestamp);
```

Comment:
nit:
```suggestion
                timestamp = createdTimestamps.getOrDefault(endpoint, System.nanoTime);
```

Reply:
This was actually intentional since I thought it would be useful to know which endpoints are in the same ramping up group. Let me know if you're strong on this though 🙏

Reply:
I didn't know that was intentional. Let's keep it as is. 😄